File: 54-SpiralMatrix/Go/main.go
/*
54. Spiral Matrix
Given an m x n matrix, return all elements of the matrix in spiral order.

Example 1:
Input: matrix = [[1,2,3],
                [4,5,6],
                [7,8,9]]
Output: [1,2,3,6,9,8,7,4,5]
Example 2:
Input: matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]]
Output: [1,2,3,4,8,12,11,10,9,5,6,7]

Constraints:
m == matrix.length
n == matrix[i].length
1 <= m, n <= 10
-100 <= matrix[i][j] <= 100
*/

package main

func spiralOrder(matrix [][]int) []int {
	var res []int
	if len(matrix) == 0 {
		return res
	}

	rows, cols := len(matrix), len(matrix[0])
	left, right, top, bottom := 0, cols-1, 0, rows-1

	for left <= right && top <= bottom {
		// traverse left
		for col := left; col <= right; col++ {
			res = append(res, matrix[top][col])
		}
		// traverse dow
		for row := top + 1; row <= bottom; row++ {
			res = append(res, matrix[row][right])
		}

		// Traverse left if required
		if left < right && top < bottom {
			for col := right - 1; col > left; col-- {
				res = append(res, matrix[bottom][col])
			}
			// Traverse up
			for row := bottom; row > top; row-- {
				res = append(res, matrix[row][left])
			}
		}
		left++
		right--
		top++
		bottom--
	}
	return res

}

File: 54-SpiralMatrix/CPP/main.cpp
/*
54. Spiral Matrix
Given an m x n matrix, return all elements of the matrix in spiral order.

Example 1:
Input: matrix = [[1,2,3],
                [4,5,6],
                [7,8,9]]
Output: [1,2,3,6,9,8,7,4,5]
Example 2:
Input: matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]]
Output: [1,2,3,4,8,12,11,10,9,5,6,7]
 
Constraints:
m == matrix.length
n == matrix[i].length
1 <= m, n <= 10
-100 <= matrix[i][j] <= 100
*/

#include<iostream>
#include<vector>
using namespace std;

vector<int> spiralOrder(vector<vector<int>>& matrix) {
    vector<int> res;
    if(matrix.empty()) return res;

    int rows = matrix.size(), cols = matrix[0].size();
    int left = 0, right = cols -1;
    int top = 0, bottom = rows -1;

    while(left <= right && top <= bottom){
        // Traverse right
        for(int col = left; col<= right; col++){
            res.push_back(matrix[top][col]);
        }

        //Traverse down
        for(int row = top+1; row <= bottom; row++){
            res.push_back(matrix[row][right]);
        }

        // Traverse left if required
        if(left < right && top <bottom){
            for( int col = right-1; col > left; col--){
                res.push_back(matrix[bottom][col]);
            }
            // Traverse up
            for( int row = bottom; row > top; row--){
                res.push_back(matrix[row][left]);
            }
        }
        left++;
        right--;
        top++;
        bottom--;
    }
    return res;
}